libuavcan/include/libuavcan/platform/memory.hpp

```diff
@@ -0,0 +1,313 @@
+     * the allocate method.
+     */
+    typename std::aligned_storage<sizeof(Block), alignof(Block)>::type storage_[NumBlocksParam];
+
+    /**
+     * The head of our free list. Note that this is a singly linked list and is unordered.
+     * If we have a pointer to a block then it is deallocated. We do not track allocations currently
+     * although we should since an application can insert data that is not in the pool by deallocating
+     * the wrong pointer.
+     */
+    std::atomic<Block*> free_list_;
+
+    StaticMemoryPool() noexcept
+        : storage_()
+        , free_list_()
+    {
+        free_list_.store(reinterpret_cast<Block*>(&storage_));
+        static_assert(NumBlocksParam > 0, "Cannot create a PoolAllocator with less than 1 block.");
+
+        Block* list = free_list_.load();
+        Block* next = &list[NumBlocksParam - 1];
+        next->next  = nullptr;
+        for (std::size_t i = NumBlocksParam - 1; i > 0; --i)
+        {
+            Block* previous = &list[i - 1];
+            previous->next  = next;
+            next            = previous;
+        }
+    }
+
+    // +------------------------------------------------------------------+
+    // | RULE OF SIX
+    // +------------------------------------------------------------------+
+    ~StaticMemoryPool() = default;
+
+public:
+    StaticMemoryPool(const StaticMemoryPool&)  = delete;
+    StaticMemoryPool(const StaticMemoryPool&&) = delete;
+    StaticMemoryPool& operator=(const StaticMemoryPool&) = delete;
+    StaticMemoryPool& operator=(const StaticMemoryPool&&) = delete;
+
+private:
+    // +------------------------------------------------------------------+
+    // | ALLOCATORS
+    // |    A list of allocators that can use this memory pool.
+    // +------------------------------------------------------------------+
+    template <std::size_t, std::uint8_t, typename, typename>
+    friend class PoolAllocator;
+
+    /**
+     * Get a reference to the static memory pool. The first time this is called
+     * the memory pool will perform internal initialization that is only thread-safe
+     * if the compiler has generated the proper thread-safe static access instructions.
+     */
+    static StaticMemoryPool& getReference()
+    {
+        static StaticMemoryPool pool_;
+        return pool_;
+    }
+
+    void* allocate(std::size_t size)
+    {
+        Block* previous = nullptr;
+        if (size <= BlockSizeParam)
+        {
+            Block* next = nullptr;
+            do
+            {
+                previous = free_list_.load();
+                if (previous == nullptr)
+                {
+#ifdef LIBUAVCAN_ENABLE_EXCEPTIONS
+                    std::bad_alloc exception;
+                    throw exception;
+#endif
+                    break;
+                }
+                next = previous->next;
+            } while (!std::atomic_compare_exchange_weak(&free_list_, &previous, next));
```

Comment:
The CAS operation has a greater chance of succeeding in the face of a bad collaborator then coarse-grained locks since the former gives every thread a stochastic chance at "winning" but the latter will deadlock other threads until the bad actor releases the lock.

That said, I agree that this could be made configurable. We can't predict what requirements a "hard realtime system" will have so we'll have to accommodate different strategies. For example, the CAS operation might be optimal for a micro-controller where the critical timing is of the interrupts. In that configuration the critical code in handler mode will have exclusive access to the memory pool so the CAS operation will always succeed (although I could change this to use `atomic_compare_exchange_strong` to make my statement more true) and the code running in thread mode will have thread-safe access to the pool without disabling interrupts which would violate the real-time constraints of this theoretical system. 

Let me propose three supported configurations we can use anywhere that resources are shared in libuavcan:

1. **no thread-safety** – No accommodations are provided since the system does not share any libuavcan resources across threads.

2. **coarse-grained locking** – Highly inefficient but if you don't have CAS operations available you have no other choice.

3. **lockless** - What I have here; CAS and other atomic operations provide thread-safety.

Reply:
Within your model of memory management, the three options seem sensible.

On a larger scope, though, I am not sure I understand the motivation for sharing a single memory pool across multiple execution contexts (threads/interrupts) unless you are planning on supporting multi-thread environments natively. Are you?

Reply:
STL requires a static memory pool since the allocator is given by type not by reference. Of course you can provide this memory pool per-thread/task since that context is also available to any given type but anything finer grained adds complexity. For microcontrollers the static patten works well since you can control the memory layout using a linker script.

As you know, malloc and free are thread-safe for most systems so the pool allocator is simply providing the same guarantees that the STL heap allocator does.

As for STL support; my main push for enabling STL allocators is to reuse std::vector and the std::priority_list adaptor instead of building custom dynamic arrays. I'm not sure if anything else in STL is worth the trouble of supporting for libuavcan but there might be some need for std::unordered_map if we need a hash map in the transport layer?

Finally, this is an iterative change. As v1 matures we'll define a default configuration for linux et-al that uses the STL heap allocator and a default configuration for no-system builds (e.g. microcontrollers) that uses this static pool allocator by default. We'll then provide a guide and perhaps other allocators to help system integrators optimize libuavcan's memory usage for their application. For example, a FreeRTOS port will probably want a per-task memory pool allocator whereas (per your previous comments) the PX4 will have other requirements we need to meet.

Reply:
> As for STL support; my main push for enabling STL allocators is to reuse std::vector and the std::priority_list adaptor instead of building custom dynamic arrays. I'm not sure if anything else in STL is worth the trouble of supporting for libuavcan but there might be some need for std::unordered_map if we need a hash map in the transport layer?

This is amazing. I understand your approach now, thanks.